keras_nlp/metrics/rouge_l.py

```diff
@@ -0,0 +1,162 @@
+            raise ValueError(
+                "`dtype` must be a floating point type. "
+                f"Received: dtype={dtype}"
+            )
+
+        if metric_type not in ("precision", "recall", "f1_score"):
+            raise ValueError(
+                '`metric_type` must be one of "precision", "recall", '
+                f'"f1_score". Received: metric_type={metric_type}'
+            )
+
+        self.metric_type = metric_type
+        self.use_stemmer = use_stemmer
+
+        # To-do: Add an option for adding custom tokenizer after the maintainers
+        # of rouge-score have released a new version.
+        self._rouge_l_scorer = rouge_scorer.RougeScorer(
+            rouge_types=["rougeL"],
+            use_stemmer=use_stemmer,
+        )
+
+        self._rouge_l_score = self.add_weight(
+            name="rouge_l_score",
+            initializer="zeros",
+            dtype=self.dtype,
+        )
+        self._number_of_samples = self.add_weight(
+            name="number_of_samples", initializer="zeros", dtype=self.dtype
+        )
+
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        # Three possible shapes for y_true and y_pred: Python string,
+        # [batch_size] and [batch_size, 1]. In the latter two cases, we have
+        # strings in the tensor/list.
+
+        # Check if input is a raw string/list.
+        if isinstance(y_true, str):
+            y_true = tf.constant([y_true])
+        elif isinstance(y_true, list):
+            y_true = tf.constant(y_true)
+        if isinstance(y_pred, str):
+            y_pred = tf.constant([y_pred])
+        elif isinstance(y_pred, list):
+            y_pred = tf.constant(y_pred)
+
+        # If the shape of y_true and y_pred is [batch_size, 1], squeeze it to
+        # [batch_size].
+        if y_true.shape.rank == 2:
+            y_true = tf.squeeze(y_true, axis=1)
+        if y_pred.shape.rank == 2:
+            y_pred = tf.squeeze(y_pred, axis=1)
+
+        batch_size = tf.shape(y_true)[0]
+
+        def _calculate_rouge_l_score(reference, hypothesis):
+            reference = tensor_to_string_list(reference)
+            hypothesis = tensor_to_string_list(hypothesis)
+            score = self._rouge_l_scorer.score(reference, hypothesis)["rougeL"]
+
+            if self.metric_type == "precision":
+                score = score.precision
+            elif self.metric_type == "recall":
+                score = score.recall
+            else:
+                score = score.fmeasure
+            return score
+
+        for batch_idx in range(batch_size):
+            score = tf.py_function(
+                func=_calculate_rouge_l_score,
+                inp=[y_true[batch_idx], y_pred[batch_idx]],
+                Tout=self.dtype,
+            )
+            self._rouge_l_score.assign_add(score)
+
+        self._number_of_samples.assign_add(
+            tf.cast(batch_size, dtype=self.dtype)
+        )
+
+    def result(self):
```

Comment:
Open question, how are these rouge scores usually aggregated when reported in a paper? We may want to do a little bit of a dive into this, to understand what the critical user journeys are when reporting an aggregate score.

Looking at the aggregation code in the package, they have a lot more there...
https://github.com/google-research/google-research/blob/master/rouge/scoring.py#L61

OK if we don't need all that, but we should make sure we understand what people will want when using this metric.

Reply:
Hmmm, good point. I went through some examples online. In particular, I went through PyTorch Ignite's ROUGE metric. Have a look: https://pytorch.org/ignite/_modules/ignite/metrics/nlp/rouge.html#Rouge (`_BaseRouge` class).

They take the average:
```
    def compute(self) -> Mapping:
        if self._num_examples == 0:
            raise NotComputableError("Rouge metric must have at least one example before be computed")

        return {
            f"{self._metric_name()}-P": float(self._precision / self._num_examples),
            f"{self._metric_name()}-R": float(self._recall / self._num_examples),
            f"{self._metric_name()}-F": float(self._fmeasure / self._num_examples),
        }
```

Reply:
Sounds good to me! We can always see if people open up issues. Thanks for checking!